src/client/api/metrics.c

```diff
@@ -0,0 +1,914 @@
+}
+
+static void
+dc_metrics_tls_aggr(dc_metrics_tls_data_t *value)
+{
+	int i;
+
+	/** Update stats */
+	dc_metrics_aggr_stats(&metrics_tls_agg.update_stat, \
+			&value->update_stat);
+
+	/** Fetch stats */
+	dc_metrics_aggr_stats(&metrics_tls_agg.fetch_stat, \
+			&value->fetch_stat);
+
+	/** io distribution by size */
+	metrics_tls_agg.ids.ids_updatesz += value->ids.ids_updatesz;
+	metrics_tls_agg.ids.ids_fetchsz += value->ids.ids_fetchsz;
+	for (i=0; i< DAOS_METRICS_IO_BYSIZE_COUNT; i++) {
+		metrics_tls_agg.ids.ids_updatecnt_bkt[i] += value->ids.ids_updatecnt_bkt[i];
+		metrics_tls_agg.ids.ids_fetchcnt_bkt[i]  += value->ids.ids_fetchcnt_bkt[i];
+	}
+
+	/** io distribution by protection type */
+	metrics_tls_agg.idp.idp_updatesz += value->idp.idp_updatesz;
+	for (i=0; i< DAOS_METRICS_IO_BYPTYPE_COUNT; i++) {
+		metrics_tls_agg.idp.idp_updatecnt_bkt[i] += value->idp.idp_updatecnt_bkt[i];
+		metrics_tls_agg.idp.idp_updatesz_bkt[i]  += value->idp.idp_updatesz_bkt[i];
+	}
+}
+
+static void
+dc_metrics_tls_cleanup(void *value)
+{
+	if (metrics_tls == NULL) {
+		goto out;
+	}
+	D_MUTEX_LOCK(&metrics_tls_lock);
+	d_list_del(&metrics_tls->list);
+	dc_metrics_tls_aggr((dc_metrics_tls_data_t *)value);
+	D_MUTEX_UNLOCK(&metrics_tls_lock);
+	free(metrics_tls);
+	metrics_tls = NULL;
+out:
+	return;
+}
+
+
+int
+dc_metrics_init()
+{
+	char *env;
+	int rc, i;
+
+	env = getenv("DAOS_METRICS");
+#if 0 /* Revisit */
+	is_metrics_enabled = ((env != NULL) && (strncmp(env, "ON", strlen("ON")) == 0));
+#else
+	if ((env != NULL) && (strncmp(env, "OFF", strlen("OFF")) == 0))
+		is_metrics_enabled = 0;
+	else
+		is_metrics_enabled = 1;
+#endif
+
+	if (is_metrics_enabled == 0)
+		return 0;
+
+	rc = dc_pool_metrics_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics for pool, rc = %d\n", rc);
+		D_GOTO(out, rc);
+	}
+	rc = dc_cont_metrics_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics for container, rc = %d\n", rc);
+		D_GOTO(out_cont, rc);
+	}
+	rc = dc_obj_metrics_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics for object, rc = %d\n", rc);
+		D_GOTO(out_obj, rc);
+	}
+	rc = dc_metrics_tls_init();
+	if (rc != 0) {
+		D_ERROR("Failed to initialize metrics TLS, rc = %d\n", rc);
+		D_GOTO(out_tls, rc);
+	}
+
+	for(i=0; i<RPDIST_ENUM_IDX_CNT; i++) {
+		rpdist_enum_idx[i] = DAOS_METRICS_IO_RPU;
+	}
+	rpdist_enum_idx[1] = DAOS_METRICS_IO_NO_PROT;
+	rpdist_enum_idx[2] = DAOS_METRICS_IO_RP2;
+	rpdist_enum_idx[3] = DAOS_METRICS_IO_RP3;
+	rpdist_enum_idx[4] = DAOS_METRICS_IO_RP4;
+	rpdist_enum_idx[6] = DAOS_METRICS_IO_RP6;
+	rpdist_enum_idx[8] = DAOS_METRICS_IO_RP8;
+	rpdist_enum_idx[12] = DAOS_METRICS_IO_RP12;
+	rpdist_enum_idx[16] = DAOS_METRICS_IO_RP16;
+	rpdist_enum_idx[32] = DAOS_METRICS_IO_RP32;
+	rpdist_enum_idx[48] = DAOS_METRICS_IO_RP48;
+	rpdist_enum_idx[64] = DAOS_METRICS_IO_RP64;
+	rpdist_enum_idx[128] = DAOS_METRICS_IO_RP128;
+
+	for(i=0; i<ECDIST_ENUM_IDX_CNT; i++) {
```

Comment:
(style)  spaces required around that '=' (ctx:VxV)
(style)  spaces required around that '<' (ctx:VxV)
(style)  space required before the open parenthesis '('